src/Symfony/Bundle/SecurityBundle/Security/FirewallConfigRegistry.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\SecurityBundle\Security;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestMatcher;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Stores firewall config objects.
+ *
+ * @author Maxime Steinhausser <maxime.steinhausser@gmail.com>
+ */
+class FirewallConfigRegistry
+{
+    private $firewallConfigs;
+    private $requestStack;
+    private $requestMatchers;
+
+    /**
+     * @param FirewallConfig[] $firewallConfigs
+     * @param RequestMatcher[] $requestMatchers Indexed by firewall name
+     * @param RequestStack     $requestStack
+     */
+    public function __construct(array $firewallConfigs = array(), array $requestMatchers = array(), RequestStack $requestStack = null)
+    {
+        $this->firewallConfigs = $firewallConfigs;
+        $this->requestStack = $requestStack;
+        $this->requestMatchers = $requestMatchers;
+    }
+
+    /**
+     * @param string $name The firewall name
+     *
+     * @return FirewallConfig|null
+     */
+    public function get($name)
+    {
+        foreach ($this->firewallConfigs as $config) {
+            if ($config->getName() === $name) {
+                return $config;
+            }
+        }
+    }
+
+    /**
+     * @return FirewallConfig|null
+     */
+    public function current()
+    {
+        if (!$this->requestStack) {
+            throw new \LogicException('Unable to get current firewall config without a RequestStack.');
+        }
+
+        return $this->fromRequest($this->requestStack->getCurrentRequest());
+    }
+
+    /**
+     * @param Request $request
+     *
+     * @return FirewallConfig|null
+     */
+    public function fromRequest(Request $request)
+    {
+        foreach ($this->firewallConfigs as $config) {
+            $requestMatcher = $this->getRequestMatcher($config);
+            if (null === $requestMatcher || $requestMatcher->matches($request)) {
+                return $config;
+            }
+        }
+    }
+
+    /**
+     * @return FirewallConfig[]
+     */
+    public function all()
+    {
+        return $this->firewallConfigs;
+    }
+
+    /**
+     * @return FirewallConfig[]
+     */
+    public function inContext($context)
+    {
+        $configs = array();
+        foreach ($this->firewallConfigs as $config) {
+            if ($config->getContext() === $context) {
+                $configs[] = $config;
+            }
+        }
+
+        return $configs;
+    }
+
+    /**
+     * @param FirewallConfig $config
+     *
+     * @return RequestMatcher|null
+     */
+    private function getRequestMatcher(FirewallConfig $config)
+    {
+        if (empty($config->getRequestMatcher())) {
```

Comment:
Note: this would be replaced by `if (null === $config->getRequestMatcher()) {` if https://github.com/symfony/symfony/pull/20589 is merged